extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/KinesisSinkP.java

```diff
@@ -0,0 +1,354 @@
+     */
+    private static final int MAX_REQUEST_SIZE_IN_BYTES = 5 * 1024 * 1024;
+
+    /**
+     * The number of Unicode characters making up keys is limited to a maximum
+     * of 256.
+     */
+    private static final int MAX_UNICODE_CHARS_IN_KEY = 256;
+
+    private final AmazonKinesisAsync kinesis;
+    private final String stream;
+
+    private final Buffer<T> buffer;
+
+    private ILogger logger;
+    private KinesisHelper helper;
+
+    private State state = State.READY_TO_SEND;
+    private long nextSendTime;
+    private Future<PutRecordsResult> sendResult;
+
+    public KinesisSinkP(
+            AmazonKinesisAsync kinesis,
+            @Nonnull String stream,
+            @Nonnull FunctionEx<T, String> keyFn,
+            @Nonnull FunctionEx<T, byte[]> valueFn
+    ) {
+        this.kinesis = kinesis;
+        this.stream = stream;
+        this.buffer = new Buffer<>(keyFn, valueFn);
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return true;
+    }
+
+    @Override
+    public void init(@Nonnull Outbox outbox, @Nonnull Context context) {
+        logger = context.logger();
+        helper = new KinesisHelper(kinesis, stream, logger);
+        helper.waitForStreamToActivate();
+    }
+
+    @Override
+    public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
+        return true; //watermark ignored
+    }
+
+    @Override
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        switch (state) {
+            case READY_TO_SEND:
+                handleReadyToSend(inbox);
+                return;
+            case SENDING_IN_PROGRESS:
+                handleSendingInProgress();
+                return;
+            default:
+                throw new RuntimeException("Programming error, unhandled state: " + state);
+        }
+
+    }
+
+    private void bufferFromInbox(@Nonnull Inbox inbox) {
+        if (buffer.isFull()) {
+            return;
+        }
+
+        while (true) {
+            T t = (T) inbox.peek();
+            if (t == null) {
+                //no more items in inbox
+                return;
+            }
+
+            boolean canBeBuffered = buffer.add(t);
+            if (canBeBuffered) {
+                inbox.remove();
+            } else {
+                //no more room in buffer
+                return;
+            }
+        }
+    }
+
+    private void handleReadyToSend(@Nonnull Inbox inbox) {
+        bufferFromInbox(inbox);
+        if (attemptToDispatchBufferContent()) {
+            state = State.SENDING_IN_PROGRESS;
+        }
+    }
+
+    private boolean attemptToDispatchBufferContent() {
+        if (buffer.isEmpty() || System.currentTimeMillis() < nextSendTime) {
+            return false;
+        }
+
+        List<PutRecordsRequestEntry> entries = buffer.content();
+        sendResult = helper.putRecordsAsync(entries);
+
+        nextSendTime = System.currentTimeMillis(); //todo: add some wait here?
```

Comment:
This is unnecessary now. It can stay in the past unless we see a reason to postpone next sending.

Reply:
I'll take care of this when I add reconnect strategy configuration/exponential backoff.